src/main/java/org/broadinstitute/hellbender/engine/PileupWalker.java

```diff
@@ -0,0 +1,180 @@
+package org.broadinstitute.hellbender.engine;
+
+import htsjdk.samtools.SAMFileHeader;
+import org.broadinstitute.barclay.argparser.Advanced;
+import org.broadinstitute.barclay.argparser.Argument;
+import org.broadinstitute.barclay.argparser.CommandLineException;
+import org.broadinstitute.hellbender.engine.filters.CountingReadFilter;
+import org.broadinstitute.hellbender.engine.filters.ReadFilter;
+import org.broadinstitute.hellbender.engine.filters.ReadFilterLibrary;
+import org.broadinstitute.hellbender.engine.filters.WellformedReadFilter;
+import org.broadinstitute.hellbender.utils.IntervalUtils;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.downsampling.DownsamplingMethod;
+import org.broadinstitute.hellbender.utils.downsampling.PositionalDownsampler;
+import org.broadinstitute.hellbender.utils.locusiterator.LocusIteratorByState;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * An AssemblyRegionWalker is a tool that processes an entire region of reads at a time, each marked as either "active"
+ * (containing possible variation) or "inactive" (not likely to contain actual variation). Tool authors must implement
+ * {@link #apply} to process each region. Authors must also implement methods providing default values for the
+ * various traversal parameters.
+ *
+ * {@link #apply} will be called once for each active AND inactive region, and it is up to the implementation how to
+ * handle/process active vs. inactive regions.
+ *
+ * Internally, the reads are loaded in chunks called read shards, which are then subdivided into active/inactive regions
+ * for processing by the tool implementation. Read shards should typically be much larger than the maximum assembly
+ * region size to achieve good performance, and should have sufficient padding on either end to avoid boundary artifacts
+ * for events near the shard boundaries.
+ *
+ * Read shards exist mainly as a proof-of-concept that we can shard the reads without introducing calling artifacts,
+ * which will be important for the Spark equivalent of this traversal.
+ */
+public abstract class PileupWalker extends GATKTool {
```

Comment:
@magicDGS You are correct.  I was mistaken about how `LocusWalker` works and will rewrite this code.  Thank you!